library/core/registry/src/main/java/org/quiltmc/qsl/registry/impl/sync/client/ClientRegistrySync.java

```diff
@@ -0,0 +1,260 @@
+			if (!optionalRegistry) {
+				LOGGER.warn("Received sync data without specifying registry!");
+			}
+			return;
+		}
+
+		var reg = currentRegistry;
+		var missingEntries = currentRegistry.quilt$applySyncMap(syncMap);
+
+		boolean disconnect = false;
+
+		if (!optionalRegistry) {
+			disconnect = checkMissing(handler, currentRegistryId, missingEntries);
+		}
+
+
+		if (!disconnect && reg == Registry.BLOCK) {
+			rebuildBlockStates();
+		} else if (!disconnect && reg == Registry.FLUID) {
+			rebuildFluidStates();
+		}
+
+		currentRegistry = null;
+		currentRegistryId = null;
+		currentCount = 0;
+		currentFlags = (byte) 0;
+		optionalRegistry = false;
+
+		syncMap = null;
+	}
+
+	public static void rebuildBlockStates() {
+		SynchronizedIdList.clear(Block.STATE_IDS);
+
+		for (var block : Registry.BLOCK) {
+			block.getStateManager().getStates().forEach(Block.STATE_IDS::add);
+		}
+	}
+
+	public static void rebuildFluidStates() {
+		SynchronizedIdList.clear(Fluid.STATE_IDS);
+
+		for (var fluid : Registry.FLUID) {
+			fluid.getStateManager().getStates().forEach(Fluid.STATE_IDS::add);
+		}
+	}
+
+	static Text getMessage(String type, String fallback) {
+		if (Language.getInstance().hasTranslation("quilt.core.registry_sync." + type)) {
+			return new TranslatableText("quilt.core.registry_sync." + type);
+		} else {
+			return new LiteralText(fallback);
+		}
+	}
+
+	public static boolean checkMissing(ClientPlayNetworkHandler handler, Identifier registry, Collection<SynchronizedRegistry.MissingEntry> missingEntries) {
+		boolean disconnect = false;
+
+		for (var entry : missingEntries) {
+			if (!RegistryFlag.isOptional(entry.flags())) {
+				disconnect = true;
+				break;
+			}
+		}
+
+		if (disconnect) {
+			sendSyncFailedPacket(handler, registry);
+			handler.getConnection().disconnect(getMessage("missing_entries", "Client registry is missing entries! Mismatched mods?"));
+			var builder = new StringBuilder("Missing entries for registry \"" + registry + "\":\n");
+
+			for (var entry : missingEntries) {
+				builder.append("\t- ").append(entry.identifier());
+				if (RegistryFlag.isOptional(entry.flags())) {
+					builder.append(" (Optional)");
+				}
+				builder.append("\n");
+			}
+
+			LOGGER.warn(builder.toString());
+		}
+
+		return disconnect;
+	}
+
+	private static void handleRestorePacket(MinecraftClient client, ClientPlayNetworkHandler handler, PacketByteBuf buf, PacketSender sender) {
+		restoreSnapshot();
+		createSnapshot();
+	}
+
+	public static void createSnapshot() {
+		for (var reg : Registry.REGISTRIES) {
+			if (reg instanceof SynchronizedRegistry registry && registry.quilt$requiresSyncing()) {
+				registry.quilt$createIdSnapshot();
+			}
+		}
+	}
+
+	public static void restoreSnapshot() {
+		for (var reg : Registry.REGISTRIES) {
+			if (reg instanceof SynchronizedRegistry registry && registry.quilt$requiresSyncing()) {
+				registry.quilt$restoreIdSnapshot();
+			}
+		}
+		ClientRegistrySync.rebuildBlockStates();
```

Comment:
I noticed that there are bits of static calls that has the class bit in (like this one) and many others without it; I'd recommend going through them all and add them back when it makes sense